src/crypto-providers/trezorCryptoProvider.ts

```diff
@@ -207,34 +193,43 @@ const TrezorCryptoProvider: () => Promise<CryptoProvider> = async () => {
     const changeAddressParams = getAddressParameters(changeOutputFiles, output.address, network)
     const tokenBundle = output.amount.type === TxTypes.AmountType.WITH_MULTIASSET
       ? prepareTokenBundle(output.amount.multiasset, false) : []
+    const datumHash = output.datumHash?.toString('hex')
 
-    if (changeAddressParams && signingMode === SigningMode.ORDINARY_TRANSACTION) {
-      return prepareChangeOutput(output.amount.coin, changeAddressParams, tokenBundle)
+    const addressParamsAllowed = [
+      SigningMode.ORDINARY_TRANSACTION, SigningMode.PLUTUS_TRANSACTION,
+    ].includes(signingMode)
+    if (changeAddressParams && addressParamsAllowed) {
+      return prepareChangeOutput(output.amount.coin, changeAddressParams, tokenBundle, datumHash)
     }
 
-    return ({
+    return {
       address: encodeAddress(output.address),
       amount: `${output.amount.coin}`,
       tokenBundle,
-    })
+      datumHash,
+    }
   }
 
   const _prepareStakeCredential = (
     stakeCredential: TxTypes.StakeCredential,
     stakeSigningFiles: HwSigningData[],
-  ): {path: string | number[]} | {scriptHash: string} => {
+    signingMode: SigningMode,
+  ): {path: BIP32Path} | {keyHash: string} | {scriptHash: string} => {
     switch (stakeCredential.type) {
       case (TxTypes.StakeCredentialType.KEY_HASH): {
         const path = findSigningPathForKeyHash(
           (stakeCredential as TxTypes.StakeCredentialKey).hash, stakeSigningFiles,
         )
-        if (!path) throw Error(Errors.MissingSigningFileForCertificateError)
-        return { path }
+        if (path) {
+          return { path }
+        }
+        if (signingMode === SigningMode.PLUTUS_TRANSACTION) {
+          return { keyHash: stakeCredential.hash.toString('hex') }
+        }
+        throw Error(Errors.MissingSigningFileForCertificateError)
```

Comment:
I think that would require introduction of hw-vendor-independent equivalent of `LedgerTypes.StakeCredentialParams` (path / key hash / script hash) that would be produced by the common function, and then translated into Trezor/Ledger types by hw-vendor-specific translation functions. I'm not sure this would decrease the complexity.